### ISSUE TITLE
www: master GUI tabs render as default vertical HTML list — <ul> missing class="tabs-nav" (regression from ac1877ab)

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#553]
> **Origin:** Customer support ticket. Suggested Fix reflects ticket agent investigation — do not derive fix from title or symptom alone.

## Summary

The user panel master GUI tabs render as a default vertical HTML bullet list (unstyled `<ul>`) on every user account. The CSS in `etc/skel/www/index.php` and `etc/skel/www/jquery.tabs.css` targets the `.tabs-nav` class, but the `<ul>` HTML element does NOT have that class. Result: no flexbox layout, no `tab.png` background sprites, no `min-width`, no `float: left`. Tabs collapse to default browser `<ul><li>` bullet-list rendering.

## Symptom

Customer sees:
- A vertical bullet list of links ("welcome", "ruTorrent", "File manager", "info", "wiki") at the top of the panel.
- No horizontal tab navigation.
- No styling, no hover states, no background graphics.
- The iframe content below (welcome.php BASIC USAGE block) renders correctly. Only the tab navigation strip at the top is broken.

## Root Cause

`etc/skel/www/index.php` around line 211:

```html
<div id="tabs">
        <ul>
<?php
foreach($frames AS $thisId => $thisFrame) {
```

The `<ul>` has no class attribute. Every tab styling rule in the codebase targets `.tabs-nav`:

- `etc/skel/www/jquery.tabs.css`: 26 selectors using `.tabs-nav`
- `etc/skel/www/jquery.tabs-ie.css`: 3 selectors using `.tabs-nav`
- `etc/skel/www/index.php` inline CSS lines 123, 124, 136, 142: 4 selectors using `.tabs-nav` (including `display: flex` for horizontal layout)

The regression was introduced by commit `ac1877ab` (Apr 5, 2026 — "fix: bundle skeleton web assets locally (Refs #423)"). That commit replaced:

```diff
-        <script src="https://static.pulsedmedia.com/jquery.tabs.pack.js"></script>
+        <script src="pmssTabs.js"></script>
```

The full `jquery.tabs.pack.js` plugin (jQuery UI tabs) auto-adds the `tabs-nav` class to the `<ul>` when `.tabs()` is invoked. The replacement minimal `pmssTabs.js` fallback (44 lines, click-handler only) does NOT add this class. The CSS therefore has nothing to match against.

## Reproduction

On any PMSS server with the post-Apr 5 skel:

1. Log into a user panel: `https://SERVER/user-USERNAME/`
2. `curl -s -u USERNAME:PASS https://SERVER/user-USERNAME/ | grep -A1 'id="tabs"'`
3. Observe `<ul>` has no class.
4. `grep -c 'tabs-nav' /etc/skel/www/index.php /etc/skel/www/jquery.tabs.css`
5. Observe all CSS targets `.tabs-nav` — selector mismatch.
6. Browser view: tabs appear as a bare HTML bullet list, vertically stacked.

Verified on a current PMSS server (PMSS git/main@2026-05-10, Debian 12.13).

## Environment

- PMSS version: git/main@2026-05-10 (current)
- Debian: 12.13
- Affected: every user account with current `index.php` skel (Apr 5+)
- Customer-visible impact: customer panels visibly "broken" since browsers' cached copies of the old CDN-loaded plugin started expiring through April-May.

## Suggested Fix

**Option A (recommended — minimal):** Hardcode the class in the HTML template:

```diff
 <div id="tabs">
-        <ul>
+        <ul class="tabs-nav">
 <?php
 foreach($frames AS $thisId => $thisFrame) {
```

Smallest possible fix. Makes the CSS apply unconditionally regardless of whether `pmssTabs.js` or any tabs plugin loads. Matches what the old `jquery.tabs.pack.js` plugin did at runtime via JS.

**Option B:** Extend `pmssTabs.js` to add the class on init:

```diff
 return this.each(function () {
     var container = $(this);
+    container.find('> ul').addClass('tabs-nav');
     var panes = container.find('> #content > .tabs-container');
```

Option A is preferred because it makes the layout work even when jQuery itself fails to load (CDN block, content blocker, network issue). Degradation is "tabs as styled horizontal list with flexbox" rather than "naked unstyled HTML bullet list."

## Risk

- **Scope:** every user on every PMSS server with the Apr 5+ skel.
- **Visibility in the wild:** intermittent because browsers cached the old CDN-loaded `jquery.tabs.pack.js` for weeks. Cache expiry exposed the regression on a rolling basis, producing scattered "panel looks broken" reports without an obvious common cause.
- **Customer impact:** panel is functionally usable (links work as fragment anchors) but visibly broken; perception is "your panel is broken." Increased support volume since Apr 5.

## Related

- ac1877ab — introduced the regression (Refs #423)
- bc86d54b — addressed a sibling issue (tabs plugin load failure → height cascade) but did not address the missing class
- GH #423 — original "bundle skeleton web assets locally" task; this is a feature-parity gap in that work

**Comment by MagnaCapax** [comment]
## Adversarial Review

**Issue:** #553 — www: master GUI tabs render as default vertical HTML list — `<ul>` missing `class="tabs-nav"` (regression from ac1877ab)
**Reviewed:** 2026-05-16
**Verdict:** PROCEED

---

### Adversarial Findings

- **[STALE_DIAGNOSIS — C]** The issue body states: "The replacement minimal `pmssTabs.js` fallback (44 lines, click-handler only) does NOT add this class." This is no longer accurate. Commit `e42cad6f` (2026-05-14, two days before this issue was filed) added `container.find('> ul').addClass('tabs-nav')` to `pmssTabs.js`. `SkeletonWebLocalAssetTest::testLocalTabsHelperAddsNavigationClass` was also added to lock in that behavior. The JS-side gap described in the issue body has already been closed. **This does not invalidate the fix** — it changes the framing from "Option A or Option B" to "Option A is the remaining gap; Option B is already done."

- **[DEFENSE-IN-DEPTH GAP — M]** Even with the `e42cad6f` JS fix in place, the `<ul>` at `index.php:289` carries no class attribute in the served HTML. If jQuery fails to load (Google CDN unreachable, content blocker, `ajax.googleapis.com` blocked at a firewall), `pmssTabs.js`'s own guard (`if (typeof $ === 'undefined') { return; }`) causes it to exit immediately — the class is never added, and all `.tabs-nav` selectors match nothing. The panel degrades to a bullet list. The HTML-hardcoded class eliminates this JS-dependency for the layout. GH #490 (bundle jQuery locally) is open but not yet merged; until it lands, any jQuery CDN failure reintroduces this failure path.

- **[RELATED OPEN ISSUE — L]** GH #536 (OPEN) targets the same root regression from `ac1877ab` but addresses only the JS layer. #536 and #553 are complementary, not duplicates. #536 should remain open and be tracked independently; #553 addresses the orthogonal HTML layer. Cross-reference both in the commit message.

- **[TEST GAP — M]** `SkeletonWebLocalAssetTest::testLocalTabsHelperAddsNavigationClass` verifies the JS side. There is no companion assertion that the HTML template itself carries `class="tabs-nav"` on the `<ul>`. A future refactor could silently drop the attribute with no test failure. The implementer should add this machine-verification as part of the same commit.

- **[SUPERSESSION CHECK — PASS]** GH #490 (bundle jQuery locally) and GH #15 (Master GUI Refactoring) are both open. #490 removes the CDN dependency that makes this defense-in-depth fix most valuable, but does not supersede it — the HTML class is correct regardless of jQuery source. #15 is long-horizon with no timeline; this one-line fix is fully justified as an interim correction. No supersession applies.

---

### Caller Analysis

Not applicable — this change ADDS a class attribute to HTML. No removal or replacement of functionality occurs.

Confirmed: no code in the repository checks for the absence of `class="tabs-nav"` on the tab `<ul>`. The attribute addition is strictly additive.

---

### Investigation Results

**Affected file:** `etc/skel/www/index.php:289`

```html
<!-- CURRENT — broken when JS unavailable -->
<div id="tabs">
        <ul>

<!-- FIX — correct with or without JS -->
<div id="tabs">
        <ul class="tabs-nav">
```

**CSS selectors requiring the class:** 26 in `etc/skel/www/jquery.tabs.css` + 3 in `etc/skel/www/jquery.tabs-ie.css` + 4 inline in `index.php` (lines 123, 124, 136, 142 — including `display: flex` for horizontal layout). All 33 selectors require `.tabs-nav` on the `<ul>` to match.

**JS path:** `pmssTabs.js` adds the class via `container.find('> ul').addClass('tabs-nav')` when `$('#tabs').tabs()` is invoked in the inline script — but only when both jQuery (Google CDN) and `pmssTabs.js` load successfully.

**Skel deployment path:** `scripts/lib/update/users/filesystem.php:189` lists `'www/index.php'` in the user sync array. The fix propagates to all existing user homedirs on next PMSS update via `pmssUserUpdateFiles()`.

**Relevant commit history for `etc/skel/www/index.php`:**
- `d1738d2a` (2026-05-16) — PHP 7.3 compat: removed typed return signatures. Does **not** touch the `<ul>` class.
- `bc86d54b` (2026-05-15) — Hardened height cascade against tabs-plugin load failure. Does **not** touch the `<ul>` class.
- `e42cad6f` (2026-05-14) — Added `addClass('tabs-nav')` to `pmssTabs.js`. Partial fix; JS-side only.
- `ac1877ab` (2026-04-05) — Introduced the regression: replaced `jquery.tabs.pack.js` (which auto-injected the class at runtime) with `pmssTabs.js` (which, at that time, did not).

**Existing tests covering this area:**
- `SkeletonWebLocalAssetTest::testIndexLocalFallbackUsesBundledTabsAssets` — verifies `pmssTabs.js` and `jquery.tabs.css` are referenced
- `SkeletonWebLocalAssetTest::testLocalTabsHelperAddsNavigationClass` — verifies `pmssTabs.js` has `addClass('tabs-nav')`
- `SkeletonWebLocalAssetTest::testUserFilesystemSyncIncludesBundledGuiAssets` — verifies skel sync includes the file
- **No test** verifies the HTML template itself carries `class="tabs-nav"` on the `<ul>`

---

### Recommendations

1. **Implement Option A only.** Option B (extend `pmssTabs.js`) was already implemented by `e42cad6f`. Do not reimplement it.

2. **Add a test.** In `SkeletonWebLocalAssetTest.php`, add:
   ```php
   public function testIndexTabsNavUlHasClassInTemplate(): void
   {
       $this->pmssAssertRepoFileContainsString(
           'etc/skel/www/index.php',
           '<ul class="tabs-nav">',
           'index.php tabs <ul> must carry class="tabs-nav" for CSS to apply without JavaScript.'
       );
   }
   ```
   This makes the HTML contract machine-verified, symmetric with the existing JS-side test.

3. **Cross-reference `e42cad6f` and #536 in the commit message.** The JS-side partial fix and this HTML-side completion are two halves of the same regression fix from `ac1877ab`.

---

### Worth-Implementing Analysis

- **Cost:** (a) Build: 1 HTML attribute + 1 test assertion, under 5 minutes. (b) Maintenance: zero recurring events — pure additive attribute, no tuning path, no state. The sole future maintenance event is removal if #15 replaces `index.php` wholesale. (c) Blast radius if it breaks: adding a class to HTML cannot break CSS that already targets that class; no regression path exists. (d) Undo cost: clean `git revert` — trivially reversible two-way door.

- **Benefit:** Every user on every server with post-Apr-5 skel gets styled horizontal tab navigation regardless of JavaScript availability — including privacy-browser users, corporate-firewall environments blocking Google CDN, and content-blocker users. Eliminates the "panel looks broken" ticket class for the JS-fail scenario.

- **Concrete present demand:** Customer ticket #111132 required 7 rounds of investigation before the root cause was confirmed by inspecting rendered HTML directly. The regression has generated scattered support load since April as browser caches of the old CDN-loaded plugin expired.

- **Verdict rationale:** PROCEED. Trivial implementation cost, zero blast radius, closes a real customer-visible regression for an entire failure class (JS unavailable). The existing JS-side fix covers the case where JS works — this completes the contract for the case where it does not.

---

### Merit Check (Tier A)

1. **Concrete present demand.** Ticket #111132 (7 rounds), and the broader series of panel-regression tickets from May 2026. Customer-quoted symptom: "tabs rendering as vertical bullet list" — fleet-wide on all post-Apr-5 skel deployments. Cache expiry pattern: browsers held the old CDN plugin for weeks post-Apr-5, producing a rolling exposure window rather than a single incident.

2. **Function.** Adds `class="tabs-nav"` to the `<ul>` at `index.php:289`, making all 33 CSS selectors across `jquery.tabs.css`, `jquery.tabs-ie.css`, and `index.php` inline styles match unconditionally, independent of JavaScript execution state.

3. **Success metric.** Post-PMSS-update: `curl -s -u USER:PASS https://SERVER/user-USER/ | grep -c 'class="tabs-nav"'` returns 1 for every user panel on every updated server. "Tabs as bullet list" customer-reported tickets return to zero within 30 days. Both signals existed before this change and are measurable independently of the code.

4. **Catastrophic if implemented.** Adding a CSS class to a `<ul>` has no realistic failure chain. The class was previously added dynamically by the jQuery UI plugin at runtime — this change moves it to static HTML. The rendered DOM state is identical in the working-JS case; this fix adds the class in the non-working-JS case where it was absent. No regression path identified.

5. **Catastrophic if NOT implemented.** Any user whose browser does not execute Google CDN jQuery (privacy extension, CDN block, regional routing failure) continues seeing bullet-list navigation → panel appears visibly broken → "your service doesn't work" perception → support ticket load → churn risk. The rolling cache expiry pattern means the affected population grows weekly until the HTML fix deploys.

6. **Deletion alternative.** Could rewrite all 33 CSS selectors to target `ul` instead of `.tabs-nav` — this requires modifying three files, breaks the existing `pmssTabs.js` `addClass` contract, and is strictly more complex. No deletion path exists; the additive HTML attribute is the minimal correct fix.

7. **Cost (four parts).** (a) 1 HTML attribute + 1 test assertion. (b) No recurring maintenance event; only removal if #15 replaces `index.php`. (c) Zero blast radius — additive change, no failure mode. (d) `git revert` — clean two-way door.

8. **Dependency blast.** `index.php` syncs to `/home/<user>/www/index.php` for all users on all servers via `pmssUserUpdateFiles()`. No external API consumers, no agents parsing the HTML structure, no scripts checking for the presence or absence of the class. Propagation is correct through the existing skel sync path.

9. **Requirements challenge.** The dumbest implicit requirement in this issue is "the CSS class must be applied by JavaScript." This was never a design intent — it was a side effect of the old jQuery UI plugin's runtime behavior that became a hidden dependency when `pmssTabs.js` replaced it. The actual requirement is "the tabs render styled." The HTML-hardcoded class satisfies this with zero complexity overhead and removes the hidden runtime dependency entirely.

---

### Anti-Pattern Hunt

- **REGEX_FOR_SEMANTICS:** PASS — not applicable
- **PREMATURE_OPTIMIZATION:** PASS — not applicable
- **OPERATOR_HANDOFF_ANTIPATTERN:** PASS — fix is complete capability closure
- **LAZY_HALT_AFTER_AUTHORIZATION:** PASS — not applicable
- **FALSE_COMPLETION_CLAIM:** PASS — the issue accurately represents an open problem; the stale pmssTabs.js claim is an investigative error, not a false completion claim
- **FABRICATED_AUTHORITY_OR_PATH:** PASS — all paths verified in source (`etc/skel/www/index.php:289`, `etc/skel/www/pmssTabs.js`, `etc/skel/www/jquery.tabs.css`)
- **FAIL_OPEN_SAFETY_BEHAVIOR:** PASS — not applicable
- **DRY_RUN_THEATER:** PASS — not applicable
- **INVESTIGATION_LOOP_THEATER:** PASS — the 7-round ticket history is documented evidence, not a proposed work pattern
- **LACK_OF_INVESTIGATION:** MINOR PASS — issue author did not verify whether `e42cad6f` had already partially addressed the root cause before filing. Counter-signals checked during this review: `git log -S "addClass"` on `pmssTabs.js` (found `e42cad6f`), current file content inspection. Both queried.

---

### Decomposition Targets

- **[REFACTOR]** `etc/skel/www/index.php:120-210` — 90-line inline `<style>` block contains `.tabs-nav` selectors that duplicate the concern owned by `jquery.tabs.css`. Two maintenance locations for the same CSS class. Low urgency; no correctness impact.

- **[DEAD CODE]** `etc/skel/www/index.php:105-145` — Remote frames `eval()` path evaluates PHP code fetched from a remote URL. `PMSS_DISABLE_REMOTE_FRAMES=1` is enforced in all deployments via `template.lighttpd` and verified by `testLighttpdTemplateDisablesRemoteFrames`. The eval path is permanently inactive in production. Candidate for removal under #15 (Master GUI Refactoring) — requires caller enumeration gate before proceeding.

- **[TEST]** `scripts/lib/tests/development/SkeletonWebLocalAssetTest.php` — Missing HTML-side contract test for `<ul class="tabs-nav">` in `index.php`. The JS-side is machine-verified; the HTML side is not. One assertion closes this gap.

— Väinämöinen

**Comment by MagnaCapax** [comment]
Another customer-reported occurrence via support ticket — same root cause, mobile-specific visual.

On iPhone mobile Safari at narrow viewport widths, the unstyled `<ul>` does not render as a recognizable bullet list — instead the default-styled list items crush into a small dark band at the top of the page and appear as overlapping/ghost text. The customer initially reported it as "menu is barely visible and text overlaps each other" rather than "tabs are a bullet list."

Same root cause (missing `class="tabs-nav"` on `<ul>`), additional mobile presentation worth noting so when the fix lands we verify both desktop bullet-list and mobile crushed-text symptoms resolve.
